applications/RomApplication/custom_strategies/rom_builder_and_solver.h

```diff
@@ -438,88 +455,179 @@ class ROMBuilderAndSolver : public BuilderAndSolver<TSparseSpace, TDenseSpace, T
-                    if(PhiElemental.size1() != dofs.size() || PhiElemental.size2() != mNumberOfRomModes) {
-                        PhiElemental.resize(dofs.size(), mNumberOfRomModes,false);
+                #pragma omp for nowait
+                for (int k = 0; k < static_cast<int>(nconditions); k++){
+                    auto it = cond_begin + k;
+
+                    // Detect if the element is active or not. If the user did not make any choice the condition is active by default
+                    bool condition_is_active = true;
+                    if ((it)->IsDefined(ACTIVE)) {
+                        condition_is_active = (it)->Is(ACTIVE);
                     }
-                    if(aux.size1() != dofs.size() || aux.size2() != mNumberOfRomModes) {
-                        aux.resize(dofs.size(), mNumberOfRomModes,false);
+
+                    // Calculate condition contribution
+                    if (condition_is_active) {
+                        Condition::DofsVectorType dofs;
+                        it->GetDofList(dofs, r_current_process_info);
+                        pScheme->CalculateSystemContributions(*it, LHS_Contribution, RHS_Contribution, EquationId, r_current_process_info);
+                        const auto &geom = it->GetGeometry();
+                        if(PhiElemental.size1() != dofs.size() || PhiElemental.size2() != mNumberOfRomModes) {
+                            PhiElemental.resize(dofs.size(), mNumberOfRomModes,false);
+                        }
+                        if(aux.size1() != dofs.size() || aux.size2() != mNumberOfRomModes) {
+                            aux.resize(dofs.size(), mNumberOfRomModes,false);
+                        }
+                        RomAuxiliaryUtilities::GetPhiElemental(PhiElemental, dofs, geom, mMapPhi);
+                        noalias(aux) = prod(LHS_Contribution, PhiElemental);
+                        for(int d = 0; d < static_cast<int>(dofs.size()); ++d){
+                            if (dofs[d]->IsFixed()==false)
+                                row(tempA,dofs[d]->EquationId()) += row(aux,d);
+                            tempb[dofs[d]->EquationId()] += RHS_Contribution(d);
+                        }
                     }
-                    RomAuxiliaryUtilities::GetPhiElemental(PhiElemental, dofs, geom, mMapPhi);
-                    noalias(aux) = prod(LHS_Contribution, PhiElemental);
-                    double h_rom_weight = it_el->GetValue(HROM_WEIGHT);
-                    noalias(tempA) += prod(trans(PhiElemental), aux) * h_rom_weight;
-                    noalias(tempb) += prod(trans(PhiElemental), RHS_Contribution) * h_rom_weight;
                 }
-            }
 
-            #pragma omp for nowait
-            for (int k = 0; k < static_cast<int>(nconditions); k++){
-                auto it = cond_begin + k;
+                #pragma omp critical
+                {
+                    noalias(Arom) += tempA;

```

Comment:
this way of exploiting parallelism is good for the galernkin build but not for the QR one. In this context it implies allocating once per thread the full rectangular matrix!

insterad of doing this, generate an array of locks and protect the write to the matrix by the proper lock, or if you prefer using atomic on all the accesses